persistence/repository/CRUDRepositoryImpl.go

```diff
@@ -0,0 +1,205 @@
+	"github.com/ditrit/badaas/persistence/models"
+	"github.com/ditrit/badaas/persistence/pagination"
+	"github.com/ditrit/badaas/services/httperrors"
+	"gorm.io/gorm"
+)
+
+// Return a database error
+func DatabaseError(message string, golangError error) httperrors.HTTPError {
+	return httperrors.NewInternalServerError("database error",
+		message,
+		golangError,
+	)
+}
+
+// Implementation of the Generic CRUD Repository
+type CRUDRepositoryImpl[T models.Tabler] struct {
+	CRUDRepository[T]
+	gormDatabase *gorm.DB
+}
+
+// Contructor of the Generic CRUD Repository
+func NewCRUDRepository[T models.Tabler](database *gorm.DB) CRUDRepository[T] {
+	return &CRUDRepositoryImpl[T]{gormDatabase: database}
+}
+
+// Run the function passed as parameter, if it returns the error and rollback the transaction.
+// If no error is returned, it commits the transaction and return the interface{} value.
+func (repo *CRUDRepositoryImpl[T]) Transaction(transactionFunction func(CRUDRepository[T]) (any, error)) (any, error) {
+	transaction := repo.gormDatabase.Begin()
+	defer func() {
+		if recoveredError := recover(); recoveredError != nil {
+			transaction.Rollback()
+		}
+	}()
+	returnValue, err := transactionFunction(&CRUDRepositoryImpl[T]{gormDatabase: transaction})
+	if err != nil {
+		transaction.Rollback()
+		return nil, err
+	}
+	return returnValue, transaction.Commit().Error
+}
+
+// Create an entity of a Model
+func (repo *CRUDRepositoryImpl[T]) Create(entity *T) httperrors.HTTPError {
+	err := repo.gormDatabase.Create(entity).Error
+	if err != nil {
+		if gormdatabase.IsDuplicateKeyError(err) {
+			fmt.Println(ErrAlreadyExists)
+			return ErrAlreadyExists
+		}
+		return DatabaseError(
+			fmt.Sprintf("could not create  %v in %s", entity, (*entity).TableName()),
+			err,
+		)
+
+	}
+	return nil
+}
+
+// Delete an entity of a Model
+func (repo *CRUDRepositoryImpl[T]) Delete(entity *T) httperrors.HTTPError {
+	err := repo.gormDatabase.Delete(entity).Error
+	if err != nil {
+		return DatabaseError(
+			fmt.Sprintf("could not delete %v in %s", entity, (*entity).TableName()),
+			err,
+		)
+	}
+	return nil
+}
+
+// Save an entity of a Model
+func (repo *CRUDRepositoryImpl[T]) Save(entity *T) httperrors.HTTPError {
+	err := repo.gormDatabase.Save(entity).Error
+	if err != nil {
+		return DatabaseError(
+			fmt.Sprintf("could not save user %v in %s", entity, (*entity).TableName()),
+			err,
+		)
+	}
+	return nil
+}
+
+// Get an entity of a Model By ID
+func (repo *CRUDRepositoryImpl[T]) GetByID(id uint) (*T, httperrors.HTTPError) {
+	var entity T
+	transaction := repo.gormDatabase.First(&entity, "id = ?", id)
+	if transaction.Error != nil {
+		return nil, DatabaseError(
+			fmt.Sprintf("could not get %s by id %d", entity.TableName(), id),
+			transaction.Error,
+		)
+	}
+	return &entity, nil
+}
+
+// Get all entities of a Model
+func (repo *CRUDRepositoryImpl[T]) GetAll(sortOps ...pagination.SortOption) ([]*T, httperrors.HTTPError) {
```

Comment:
rename `sortOps` to something better like `sortOptions` and do the same for `ops`